### PR TITLE
Unequal partitioning through partitioning file

### PR DESCRIPTION
**Pull request by suratdas**
In the image below, we wanted Windows, Mac and Linux cases to run in parallel as defined in the partitioning file. This could not be done with default partitioning which works by splitting the test cases equally. This pull request implements suite based unequal split. 

**Partitioning file**
```
Page	Partition	Test System	Order
Demo.DemoSuite.Linux	0	slim	0
Demo.DemoSuite.Mac	1	slim	0
Demo.DemoSuite.Windows	2	slim	0
```

image

### CODE REVIEW

**Comment by fhoeben** [comment]
I completely understand splitting equally is not always what one wants (in my own project I create a partitioner that does something similar, and generates the split based on previous run times). I'll look at your changes this weekend.

As a side note: for your specific case of running 3 suites in parallel wouldn't it be easier just to start one run per suite? Then there is no need to work based on a partitioning file.

**Comment by fhoeben** [comment]
Changes look pretty minimal and logical. Could you also add some (unit) tests?

**Comment by suratdas** [comment]
I presented the case in a simplified way but we sometimes have more than one suite in a partition. We change and decide how many suites to go in which partition depending on priority/test environment/execution time etc.. It was easier to manage via partition file rather than modifying Jenkins parameters for run per suite.

> As a side note: for your specific case of running 3 suites in parallel wouldn't it be easier just to start one run per suite? Then there is no need to work based on a partitioning file.

**Comment by fhoeben** [comment]
Do I read your change correctly that the existing code already allowed for uneven split, your change just makes for simpler configuration by not requiring each page to be mentioned. Instead one can configure suites to be placed in a partition. 

Configuration per test page is checked first, then any configuration up the tree is checked and only then does the equal partitioning kick in. Correct?

Seems perfectly fine, but it does seem to warrant some test cases. Will you be adding those?

**Comment by suratdas** [comment]
Yes, you are absolutely correct.
Right now, I am going through the tests to understand and then will add unit tests.

**Comment by suratdas** [comment]
Unit tests added and documentation updated.

File: src/fitnesse/util/partitioner/MapBasedListPartitioner.java [hunk 1]
```diff
@@ -21,7 +23,8 @@ public MapBasedListPartitioner(Function<T, String> keyFunction, Map<String, Inte
                                  BiFunction<List<List<T>>, List<T>, List<List<T>>> notFoundFunction) {
     super(
       t -> {
-        String key = keyFunction.apply(t);
+        Object object = keyFunction.apply(t);
```
Comment (fhoeben) [inline src/fitnesse/util/partitioner/MapBasedListPartitioner.java:26]: Is this change needed? It doesn't look like you made any changes to production code that require this.
If it is necessary the type of the parameter also needs to change. If you just need it for tests you can probably incorporate the handling of `WikiPage` results into the function you pass in, so this code always gets a `String` from the keyFunction

**Review by fhoeben — CHANGES_REQUESTED** [review]
Looks quite good. Almost there.

File: src/fitnesse/util/partitioner/MapBasedListPartitioner.java (hunk 1 above)
Comment (suratdas) [reply src/fitnesse/util/partitioner/MapBasedListPartitioner.java]: Thanks for the suggestion. I incorporated it now. Please review.
